File: Boop/Boop/bridges/python/bridge.py
import importlib.util
import json
import os
import sys
import urllib.request

state = json.loads(os.environ.get('BOOP_STATE', '{}'))
module_ext = os.environ.get('BOOP_MODULE_EXT', '.py')
script_dir = os.environ.get('BOOP_SCRIPT_DIR', '')
lib_dir = os.environ.get('BOOP_LIB_DIR', '')

_loaded_modules = {}

def boop_require(path):
    if not path.endswith(module_ext):
        path += module_ext
    if path.startswith('@boop/'):
        mod_path = os.path.join(lib_dir, path[6:])
    else:
        mod_path = os.path.join(script_dir, path)
    if not os.path.isfile(mod_path):
        return None
    if mod_path in _loaded_modules:
        return _loaded_modules[mod_path]
    spec = importlib.util.spec_from_file_location('boop_mod_' + str(len(_loaded_modules)), mod_path)
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    _loaded_modules[mod_path] = module
    return module

class State:
    def __init__(self, data):
        self.text = data.get('text')
        self.fullText = data.get('fullText')
        self.selection = data.get('selection')
        self.network = data.get('network', False)
        self.inserts = []
        self.messages = []
    def post_info(self, msg):
        self.messages.append({'type': 'info', 'message': msg})
    def post_error(self, msg):
        self.messages.append({'type': 'error', 'message': msg})
    def insert(self, value):
        self.inserts.append(value)
    def fetch(self, url, method=None, body=None):
        if not self.network:
            self.post_error('Network permission required')
            return None
        req = urllib.request.Request(url, data=body.encode('utf-8') if body else None, method=method or 'GET')
        try:
            with urllib.request.urlopen(req) as f:
                return f.read().decode('utf-8')
        except Exception:
            self.post_error('Failed to fetch')
            return None

state_obj = State(state)

globals()['boop_require'] = boop_require

script_path = sys.argv[1]

spec = importlib.util.spec_from_file_location('user_script', script_path)
module = importlib.util.module_from_spec(spec)
spec.loader.exec_module(module)
if hasattr(module, 'main'):
    module.main(state_obj)

output = {
    'text': state_obj.text,
    'fullText': state_obj.fullText,
    'selection': state_obj.selection,
    'inserts': state_obj.inserts,
    'messages': state_obj.messages,
}
print(json.dumps(output))

File: Boop/Boop/bridges/node/bridge.js
#!/usr/bin/env node
const {execFileSync} = require('child_process');
const fs = require('fs');
const Module = require('module');
const path = require('path');

const state = JSON.parse(process.env.BOOP_STATE || '{}');
const MODULE_EXT = process.env.BOOP_MODULE_EXT || '.njs';
const SCRIPT_DIR = process.env.BOOP_SCRIPT_DIR || '';
const LIB_DIR = process.env.BOOP_LIB_DIR || '';

require.extensions['.njs'] = require.extensions['.js'];

const originalRequire = Module.prototype.require;
Module.prototype.require = function(name) {
  let target = name;
  if(!target.endsWith(MODULE_EXT)) target += MODULE_EXT;
  let file;
  if(target.startsWith('@boop/')) {
    file = path.join(LIB_DIR, target.slice(6));
  } else {
    file = path.join(SCRIPT_DIR, target);
  }
  if(fs.existsSync(file)) return originalRequire.call(this, file);
  return originalRequire.call(this, name);
};

class State {
  constructor(data){
    this.text = data.text;
    this.fullText = data.fullText;
    this.selection = data.selection;
    this.network = data.network;
    this.inserts = [];
    this.messages = [];
  }
  post_info(msg){ this.messages.push({type:'info', message: msg}); }
  post_error(msg){ this.messages.push({type:'error', message: msg}); }
  insert(val){ this.inserts.push(val); }
  fetch(url, method, body){
    if(!this.network){ this.post_error('Network permission required'); return null; }
    try{
      const args = ['-sL'];
      if(method && method !== 'GET') args.push('-X', method);
      if(body) args.push('--data', body);
      args.push(url);
      return execFileSync('curl', args, {encoding: 'utf8'});
    }catch(e){
      this.post_error('Failed to fetch');
      return null;
    }
  }
}

const scriptPath = process.argv[2];
const stateObj = new State(state);
const user = require(path.resolve(scriptPath));
if(typeof user.main === 'function'){
  const res = user.main(stateObj);
  if(res && typeof res.then === 'function') {
    res.then(()=>finish()).catch(()=>finish());
  } else {
    finish();
  }
} else {
  finish();
}

function finish(){
  const output = {
    text: stateObj.text,
    fullText: stateObj.fullText,
    selection: stateObj.selection,
    inserts: stateObj.inserts,
    messages: stateObj.messages
  };
  console.log(JSON.stringify(output));
}
